### PR TITLE
mpi: 1D contiguous alltoall in Impl

### PR DESCRIPTION
**Pull request by cwpearson**
Just in `Impl::` for now

### CODE REVIEW

**Comment by cwpearson** [comment]
Needs documentation

**Comment by aprokop** [comment]
Needs `MPI_IN_PLACE` version.

**Comment by cwpearson** [comment]
Okay we have an in-place version too now

File: src/impl/KokkosComm_alltoall.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      ss << "alltoall sendCount * communicator size (" << sendCount << " * " << size
+         << ") is greater than send view size";
+      throw std::runtime_error(ss.str());
+    }
+    if (recvCount * size > RT::extent(rv, 0)) {
+      std::stringstream ss;
+      ss << "alltoall recvCount * communicator size (" << recvCount << " * " << size
+         << ") is greater than recv view size";
+      throw std::runtime_error(ss.str());
+    }
```
Comment (cedricchevalier19) [inline src/impl/KokkosComm_alltoall.hpp:77]: Do we want to always run these tests? Perhaps they can be conditionally compiled?

File: unit_tests/test_alltoall.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      sv.extent(0), KOKKOS_LAMBDA(const int i) { sv(i) = rank + i; });
+
+  KokkosComm::Impl::alltoall(Kokkos::DefaultExecutionSpace(), sv, nContrib, rv, nContrib, MPI_COMM_WORLD);
+
+  int errs;
+  Kokkos::parallel_reduce(
+      rv.extent(0),
+      KOKKOS_LAMBDA(const int &i, int &lsum) {
+        const int src = i / nContrib;                      // who sent this data
+        const int j   = rank * nContrib + (i % nContrib);  // what index i was at the source
```
Comment (cedricchevalier19) [inline unit_tests/test_alltoall.cpp:55]:
```suggestion
        const int j   = i - rank * nContrib;  // what index i was at the source
```

File: unit_tests/test_alltoall.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      rv.extent(0), KOKKOS_LAMBDA(const int i) { rv(i) = rank + i; });
+
+  KokkosComm::Impl::alltoall(Kokkos::DefaultExecutionSpace(), rv, nContrib, MPI_COMM_WORLD);
+
+  int errs;
+  Kokkos::parallel_reduce(
+      rv.extent(0),
+      KOKKOS_LAMBDA(const int &i, int &lsum) {
+        const int src = i / nContrib;                      // who sent this data
+        const int j   = rank * nContrib + (i % nContrib);  // what index i was at the source
```
Comment (cedricchevalier19) [inline unit_tests/test_alltoall.cpp:84]:
```suggestion
        const int j   = i - rank * nContrib;  // what index i was at the source
```

**Review by cedricchevalier19 — DISMISSED** [review]
Just a small suggestion for getting rid of `%` operator.

File: src/impl/KokkosComm_alltoall.hpp (hunk 1 above)
Comment (cwpearson) [reply src/impl/KokkosComm_alltoall.hpp:77]: Yeah I think it's a good idea to skip these in a release build.

File: unit_tests/test_alltoall.cpp (hunk 2 above)
Comment (cwpearson) [reply unit_tests/test_alltoall.cpp:55]: Consider `i=0, rank=1, nContrib=1`, previously yeilds `1`, your suggestion yields `-1`

File: src/impl/KokkosComm_alltoall.hpp (hunk 1 above)
Comment (cwpearson) [reply src/impl/KokkosComm_alltoall.hpp:77]: I want to build a KOKKOSCOMM_DEBUG definition into the `KokkosComm_config.hpp` file you will introduce. I will open an issue to track that.

**Review by cedricchevalier19 — APPROVED** [review]

**Pull request merged**